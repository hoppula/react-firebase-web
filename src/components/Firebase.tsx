import * as firebase from "firebase"
import * as PropTypes from "prop-types"
import * as React from "react"

export interface FirebaseProps {
  readonly apiKey: string
  readonly projectId: string
  readonly databaseURL?: string
}

export class Firebase extends React.Component<FirebaseProps, {}> {
  static propTypes = {
    apiKey: PropTypes.string.isRequired,
    databaseURL: PropTypes.string,
    projectId: PropTypes.string.isRequired
  }

  static childContextTypes = {
    firebase: PropTypes.object
  }

  firebase: firebase.app.App

  constructor(props: FirebaseProps) {
    super(props)
    const { apiKey, projectId, databaseURL } = props
    this.firebase = firebase.initializeApp({
      apiKey,
      authDomain: `${projectId}.firebaseapp.com`,
      databaseURL: databaseURL
        ? databaseURL
        : `https://${projectId}.firebaseio.com`,
      storageBucket: `${projectId}.appspot.com`
    })
  }

  getChildContext() {
    return {
      firebase: this.firebase
    }
  }

  render() {
    return React.Children.only(this.props.children)
  }
}

export default Firebase
