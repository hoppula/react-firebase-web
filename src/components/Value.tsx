import * as firebase from "firebase"
import * as PropTypes from "prop-types"
import * as React from "react"

import { FirebaseCallback, FirebaseContext } from "../types"

export interface ValueProps {
  readonly once?: boolean
  readonly path: string
  readonly children: (value: any) => React.ReactElement<any>
}

export interface ValueState {
  value: any
}

export class Value extends React.Component<ValueProps, ValueState> {
  static propTypes = {
    once: PropTypes.bool,
    path: PropTypes.string.isRequired,
    children: PropTypes.func.isRequired
  }

  static contextTypes = {
    firebase: PropTypes.object.isRequired
  }

  static defaultProps = {
    once: false
  }

  context: FirebaseContext
  private ref: firebase.database.Reference
  private listeners: {
    value: FirebaseCallback
  }

  constructor(props: ValueProps) {
    super(props)
    this.state = {
      value: {}
    }
  }

  componentDidMount() {
    const { path, once } = this.props
    this.ref = this.createRef(path)
    this.listeners = {
      value: null
    }
    this.addListeners(this.ref, once)
  }

  createRef(path: string) {
    return this.context.firebase.database().ref(path)
  }

  addListeners(ref: firebase.database.Reference, once: boolean) {
    const listeners = this.listeners

    // listen only once if once prop has been passed
    if (once) {
      return ref.once("value", snapshot => {
        this.setState({
          value: snapshot.val()
        })
      })
    }

    // use live bindings as default
    listeners.value = ref.on("value", snapshot => {
      this.setState(() => {
        return { value: snapshot.val() }
      })
    })
  }

  removeListeners() {
    Object.keys(this.listeners).forEach((eventType: "value") => {
      this.ref.off(eventType, this.listeners[eventType])
    })
    this.listeners = {
      value: null
    }
  }

  componentDidUpdate(prevProps: ValueProps) {
    const { path, once } = this.props
    const { path: prevPath, once: prevOnce } = prevProps
    if (path !== prevPath || once !== prevOnce) {
      this.removeListeners()
      this.ref = this.createRef(path)
      this.addListeners(this.ref, once)
    }
  }

  componentWillUnmount() {
    this.removeListeners()
  }

  render() {
    return React.Children.only(this.props.children(this.state.value))
  }
}

export default Value
