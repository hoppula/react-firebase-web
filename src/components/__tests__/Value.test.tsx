import React from "react"
import renderer from "react-test-renderer"
import { Firebase } from "../Firebase"
import Value from "../Value"

import { initializeMockSDK } from "./setup"
const mocksdk = initializeMockSDK()

jest.mock("firebase/app", () => {
  return {
    initializeApp: (...args: any[]) => mocksdk.initializeApp(...args)
  }
})

describe("Value", () => {
  test("renders most recent value whenever value changes", () => {
    let component = renderer.create(
      <Firebase apiKey="firebase-test" projectId="firebase-test">
        <Value
          path="value"
          children={value => <div>{value && value.name}</div>}
        />
      </Firebase>
    )
    const testRef = component
      .getInstance()
      .firebase.database()
      .ref("value")
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()

    testRef.set({ name: "Name 1" })
    testRef.flush()
    tree = component.toJSON()
    expect(tree).toMatchSnapshot()

    testRef.set({ name: "Name 2" })
    testRef.flush()
    tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
