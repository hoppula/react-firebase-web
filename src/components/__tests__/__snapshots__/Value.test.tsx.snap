// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Value renders most recent value whenever value changes 1`] = `<div />`;

exports[`Value renders most recent value whenever value changes 2`] = `
<div>
  Name 1
</div>
`;

exports[`Value renders most recent value whenever value changes 3`] = `
<div>
  Name 2
</div>
`;
